File: backend/internal/service/exercise_metric.go
package service

import "strings"

// Personal records were originally weight-only: a set with no external load was
// dropped before it ever reached the database, so calisthenics students (شنا،
// بارفیکس، پلانک، اسکات با وزن بدن) could never set a PR. Every logged set now
// carries a metric kind that says what "better" means for that movement, and
// records are compared only against the same kind for the same exercise.
const (
	// MetricKindWeight — external load. Best = heaviest kilo lifted.
	MetricKindWeight = "weight"
	// MetricKindReps — bodyweight movement. Best = most reps in a single set.
	MetricKindReps = "reps"
	// MetricKindHold — isometric hold. Best = longest hold in seconds.
	MetricKindHold = "hold"
)

// ValidMetricKind reports whether kind is one the API accepts.
func ValidMetricKind(kind string) bool {
	switch kind {
	case MetricKindWeight, MetricKindReps, MetricKindHold:
		return true
	default:
		return false
	}
}

// isometricNameHints are movements measured in seconds rather than reps. Both
// the Persian catalog names and their English twins are listed, because a
// coach-authored program can carry either.
var isometricNameHints = []string{
	"پلانک", "plank",
	"وال سیت", "wall sit",
	"هالو هولد", "hollow hold",
	"ال‌سیت", "ال سیت", "l-sit", "l sit",
	"وی‌سیت", "وی سیت", "v-sit",
	"بالانس روی دست", "handstand",
	"پلانچ", "planche",
	"فرانت لِوِر", "فرانت لور", "front lever",
	"بک لِوِر", "بک لور", "back lever",
	"آویزان مرده", "dead hang",
	"ایزومتریک", "isometric",
	"نگه‌داشتن", "نگه داشتن", "hold",
	"صلیب آهنین", "iron cross",
	"پرچم", "flag",
	"ابوالهول", "sphinx",
}

// bodyweightEquipment are the catalog equipment values that mean "no external
// load", in both catalog languages.
var bodyweightEquipment = map[string]bool{
	"وزن بدن":         true,
	"با وزن بدن":      true,
	"body weight":     true,
	"bodyweight":      true,
	"با کمک دستگاه":   true,
	"assisted":        true,
}

// bodyweightNameHints catch coach-authored entries that are not linked to a
// catalog row, so no equipment field is available.
var bodyweightNameHints = []string{
	"شنا", "push-up", "push up", "pushup",
	"بارفیکس", "pull-up", "pull up", "chin-up", "chin up",
	"دیپ", "dip",
	"دراز و نشست", "sit-up", "sit up", "crunch", "کرانچ",
	"برپی", "burpee",
	"جامپینگ جک", "jumping jack",
	"کوهنورد", "mountain climber",
	"لانژ", "lunge",
	"ماسل‌آپ", "ماسل آپ", "muscle-up", "muscle up",
	"زیربغل وارونه", "inverted row",
	"اینچ‌ورم", "inchworm",
	"سوپرمن", "superman",
	"دد باگ", "dead bug",
	"پل سرینی", "glute bridge",
}

// DetectMetricKind picks the metric for a logged set. `equipment` is the
// catalog value when the set is linked to a catalog exercise (may be empty);
// `hasExternalLoad` says whether the user actually entered a weight, which
// wins over everything (a weighted pull-up is a weight PR).
func DetectMetricKind(exerciseName, equipment string, hasExternalLoad bool) string {
	name := strings.ToLower(strings.TrimSpace(exerciseName))

	if containsAny(name, isometricNameHints) {
		return MetricKindHold
	}
	if hasExternalLoad {
		return MetricKindWeight
	}
	if bodyweightEquipment[strings.ToLower(strings.TrimSpace(equipment))] {
		return MetricKindReps
	}
	if containsAny(name, bodyweightNameHints) {
		return MetricKindReps
	}
	// Unknown movement with no weight entered: treat it as a rep-based effort
	// rather than discarding the set, which is what used to happen.
	return MetricKindReps
}

func containsAny(haystack string, needles []string) bool {
	for _, needle := range needles {
		if needle != "" && strings.Contains(haystack, needle) {
			return true
		}
	}
	return false
}

File: frontend/src/lib/workout/exerciseMetric.js
// Personal records used to be weight-only, so the "ثبت وزنه‌ها" panel dropped
// any exercise the student did at bodyweight. Every logged set now carries a
// metric kind saying what "better" means for that movement; this mirrors
// backend/internal/service/exercise_metric.go so the client can pick the right
// input and the right label without a round-trip.

export const METRIC_WEIGHT = "weight";
export const METRIC_REPS = "reps";
export const METRIC_HOLD = "hold";

const ISOMETRIC_HINTS = [
  "پلانک", "plank",
  "وال سیت", "wall sit",
  "هالو هولد", "hollow hold",
  "ال‌سیت", "ال سیت", "l-sit", "l sit",
  "وی‌سیت", "وی سیت", "v-sit",
  "بالانس روی دست", "handstand",
  "پلانچ", "planche",
  "فرانت لِوِر", "فرانت لور", "front lever",
  "بک لِوِر", "بک لور", "back lever",
  "آویزان مرده", "dead hang",
  "ایزومتریک", "isometric",
  "نگه‌داشتن", "نگه داشتن",
  "صلیب آهنین", "iron cross",
  "پرچم", "flag",
  "ابوالهول", "sphinx",
];

const BODYWEIGHT_EQUIPMENT = new Set([
  "وزن بدن",
  "با وزن بدن",
  "body weight",
  "bodyweight",
  "با کمک دستگاه",
  "assisted",
]);

const BODYWEIGHT_HINTS = [
  "شنا", "push-up", "push up", "pushup",
  "بارفیکس", "pull-up", "pull up", "chin-up", "chin up",
  "دیپ", "dip",
  "دراز و نشست", "sit-up", "sit up", "crunch", "کرانچ",
  "برپی", "burpee",
  "جامپینگ جک", "jumping jack",
  "کوهنورد", "mountain climber",
  "لانژ", "lunge",
  "ماسل‌آپ", "ماسل آپ", "muscle-up", "muscle up",
  "زیربغل وارونه", "inverted row",
  "اینچ‌ورم", "inchworm",
  "سوپرمن", "superman",
  "دد باگ", "dead bug",
  "پل سرینی", "glute bridge",
];

function containsAny(haystack, needles) {
  return needles.some((n) => n && haystack.includes(n));
}

/**
 * Pick the metric an exercise should be logged with.
 * @param {{name?: string, equipment?: string}} exercise
 * @returns {"weight"|"reps"|"hold"}
 */
export function detectMetricKind(exercise) {
  const name = String(exercise?.name ?? "").trim().toLowerCase();
  const equipment = String(exercise?.equipment ?? "").trim().toLowerCase();

  if (containsAny(name, ISOMETRIC_HINTS)) return METRIC_HOLD;
  if (BODYWEIGHT_EQUIPMENT.has(equipment)) return METRIC_REPS;
  if (containsAny(name, BODYWEIGHT_HINTS)) return METRIC_REPS;
  // Anything else is assumed to take external load; the user can still switch.
  return METRIC_WEIGHT;
}

export const METRIC_LABELS = {
  [METRIC_WEIGHT]: { input: "وزنه", unit: "کیلوگرم", short: "kg" },
  [METRIC_REPS]: { input: "تکرار", unit: "تکرار", short: "تکرار" },
  [METRIC_HOLD]: { input: "زمان", unit: "ثانیه", short: "ثانیه" },
};

/** Human-readable Persian summary of one record, e.g. "۳۰ تکرار". */
export function formatMetricValue(kind, { weightKg, reps, holdSeconds } = {}) {
  const fa = (n) => Number(n || 0).toLocaleString("fa-IR");
  switch (kind) {
    case METRIC_REPS:
      return `${fa(reps)} تکرار`;
    case METRIC_HOLD:
      if (holdSeconds >= 60 && holdSeconds % 60 === 0) {
        return `${fa(holdSeconds / 60)} دقیقه`;
      }
      return `${fa(holdSeconds)} ثانیه`;
    default:
      return reps > 0
        ? `${fa(weightKg)} کیلوگرم × ${fa(reps)}`
        : `${fa(weightKg)} کیلوگرم`;
  }
}

/** The number that has to go up for a record of this kind. */
export function metricValue(kind, record) {
  switch (kind) {
    case METRIC_REPS:
      return Number(record?.reps || 0);
    case METRIC_HOLD:
      return Number(record?.holdSeconds || 0);
    default:
      return Number(record?.weightKg || 0);
  }
}

/** The previous best that came with a record of this kind. */
export function previousBestValue(kind, record) {
  switch (kind) {
    case METRIC_REPS:
      return Number(record?.previousBestReps || 0);
    case METRIC_HOLD:
      return Number(record?.previousBestHoldSec || 0);
    default:
      return Number(record?.previousBestKg || 0);
  }
}

/**
 * Turn the per-exercise entries the log panel collected into the API `sets`
 * payload, dropping exercises with nothing entered.
 * @param {Array<{name: string, exerciseId?: number, equipment?: string, reps?: string}>} exercises
 * @param {Record<string, {kind: string, value: string}>} entries keyed by exercise name
 */
export function buildSetsPayload(exercises, entries) {
  return (exercises || [])
    .map((ex) => {
      const entry = entries?.[ex.name];
      const value = parseFloat(entry?.value);
      if (!Number.isFinite(value) || value <= 0) return null;
      const kind = entry?.kind || detectMetricKind(ex);
      const set = {
        exerciseName: ex.name,
        exerciseId: ex.exerciseId || undefined,
        equipment: ex.equipment || undefined,
        metricKind: kind,
        setNumber: 1,
      };
      if (kind === METRIC_REPS) {
        set.reps = Math.round(value);
      } else if (kind === METRIC_HOLD) {
        set.holdSeconds = Math.round(value);
      } else {
        set.weightKg = value;
        set.reps = firstInt(ex.reps);
      }
      return set;
    })
    .filter(Boolean);
}

/** Extract the leading integer from a prescribed rep string like "12" or "8-10". */
export function firstInt(value) {
  const m = String(value ?? "").match(/\d+/);
  return m ? parseInt(m[0], 10) : 0;
}
